crates/hir-ty/src/utils.rs

```diff
@@ -270,64 +270,129 @@ pub(crate) struct Generics {
 }
 
 impl Generics {
```

Comment:
The changes in this module are not too great perf wise, instead of borrowing we do a lot of cloning now I wonder if we can get around this somehow.

Reply:
Is there any specific reason why there is separate arena for lifetimes in `GenericParams`, if we could combine both. It could potential solve this as well having ordering of lifetimes issue.

Reply:
I don't think it would solve the ordering issues. I think you could just add something like this:
```rust
enum GenericParamDataRef<'a> {
    TypeParamData(&'a TypeParamData),
    ConstParamData(&'a ConstParamData),
    LifetimeParamData(&'a LifetimeParamData),
}
```

Reply:
I wasn't sure whether I could a add new enum since there is type with exact same data.

I was thinking that could solve the issue but now that I think about it, It's not going to solve that ordering issues and will also have other problematic things.
But what do you think about adding a flag for the presence of Self type and generating the iterator with Self at the start and then other lifetimes and types if the flag is set?